libraries/tickfs/SPEC.md

```diff
@@ -0,0 +1,494 @@
+## Storage Format
+
+TickFS flash storage is formatted as a sorting of regions with each region
+containing TickFS objects.
+
+### TickFS Regions
+
+A TickFS region is the smallest set of contiguous flash storage space that can
+be erased with a single command.
+
+For example if a flash controller allows erasing lengths of 1024 (0x400) bytes
+then the length of a region will be 1024 (0x400) bytes.
+
+The number of regions is determined by the capacity of the flash storage and
+the size of regions.
+
+The start and end address of flash used for TickFS must be region aligned.
+
+### TickFS Objects
+
+A TickFS object is the representation of a key/value pair in flash. An object
+contains the value to be saved as well as useful header data.
+
+TickFS saves and reads objects from flash. TickFS objects contain the value
+the user wanted to store as well as extra header data.
+
+A TickFS object consists of multiple parts:
+
+```
+|||||||||||||||||
+|               |
+| Object Header |
+|               |
+|||||||||||||||||
+|               |
+|     Value     |
+|               |
+|||||||||||||||||
+|               |
+|   Check Sum   |
+|               |
+|||||||||||||||||
+```
+
+#### Object Header
+
+Currently ObjectHeader includes these fields:
+
+```Rust
+struct ObjectHeader {
+    version: u8,
+    flags: u4,
+    len: u12,
+    hashed_key: u64,
+}
+```
+
+The `version` field is a byte containing the version of TickFS used when the
+object was written.
+This allows us to upgrade this library in the future, while still supporting
+old data formats.
+
+The `flags` field is a bitmap of at most 4 flags that can be OR-ed together to
+describe an object state or features. The only flag defined is the `valid` flag
+(bit 3), indicating that an object is valid.
+
+It looks like this in flash:
+
+```
+|valid|Reserved|Reserved|Reserved|
+|     |        |        |        |
+|  1  |    0   |    0   |    0   |
+```
+
+Where `valid` indicates if an object is valid. A `1` indicates it is a valid
+object, a `0` indicates that it has been marked as invalid (see below).
+
+The `len` field is 12-bits long.
+This field indicates the total length of the object, including the
+header and check sum. The maximum length of the entire object is
+4KiB (0xFFF) or the region size, whichever is smaller.
+
+The `hashed_key` field stores the 64-bit (8 byte) output of the key hash.
+
+ObjectHeader is internal to TickFS and users of TickFS do not need to
+understand it.
+
+#### Object Value
+
+The Value component of the TickFS object is the value that the user wants to
+store.
+
+The values can be any length as long as they follow both:
+ * Don't span multiple regions. That limits the maximum value length to
+   `region_size - size_of::<ObjectHeader>()`
+ * Don't have a maximum length greater then 4KiB (0xFFF).
+
+#### Check Sum
+
+The check sum is a hash of the entire object (not including the check sum).
+The check sum is calculated using the same hash algorithm used for the keys.
```

Comment:
Then it's not a checksum.

Please note that hashes (especially cryptographic hashes) and CRCs have very different error detection behaviors. A CRC (depending on the generator polynomial) provides guarantees on certain errors it can detect (e.g., an odd number of bit errors). A hash provides no guarantees. A strong cryptographic hash, for example, means that flipping one bit in the input makes every bit in the output effectively random. This means with N bits there is a 2^-N probability you get a collision. It can be OK to make this tradeoff, but a CRC is actually in some ways better.

Reply:
Yep, I thought about this.

I would prefer to use a CRC here, the problem is we then also need a CRC implementation as well as a hash implementation in the kernel. The hope is that the hash implementation can be offloaded to hardware, so re-using it for a checksum will hopefully speed things up.

I think I'll move toward using a CRC though, as I agree it will provide much more useful error checking.